### PR TITLE
Add start method import

### PR DESCRIPTION
**Pull request by karthiknadig**

### CODE REVIEW

File: ptvsd/wrapper.py [hunk 1]
```diff
@@ -2427,6 +2427,8 @@ def on_pydevd_thread_create(self, seq, args):
         # as well.
         with self.is_process_created_lock:
             if not self.is_process_created:
+                if not debugger_attached.isSet():
+                    return
```
Comment (karthiknadig) [inline ptvsd/wrapper.py]: @int19h Need your input on this. Basically i was running into a race in tests where we would get `process` message before we even send `initialize` request. The process message came from this location. I think we should not send this event unless debugger attached.

Comment (int19h) [reply ptvsd/wrapper.py]: Agreed.

File: pytests/func/test_run.py [hunk 2]
```diff
@@ -12,24 +12,37 @@
...
-@pytest.mark.parametrize('start_method', [START_METHOD_LAUNCH, START_METHOD_CMDLINE])
+@pytest.mark.parametrize('start_method', [START_METHOD_LAUNCH, START_METHOD_CMDLINE, START_METHOD_IMPORT])
 def test_run(debug_session, pyfile, run_as, start_method):
     @pyfile
     def code_to_debug():
         import os
         import sys
         import backchannel
 
+        if os.getenv('PTVSD_ENABLE_ATTACH', False):
```
Comment (int19h) [inline pytests/func/test_run.py]: This should probably go into `DebugSession.initialize()`, so we can run other tests with import too without repeating it every time. The session can generate another temp script with `@pyfile` that does the import, and then invokes `target[1]` via the `runpy` module if `target[0]` is "file" or "module". For "code", we can just prepend "import ptvsd; ...;" directly to `target[1]`.

Comment (karthiknadig) [reply pytests/func/test_run.py]: That is a good idea, will update.

File: pytests/helpers/session.py [hunk 3]
```diff
@@ -128,33 +128,39 @@ def stop(self):
...
-            argv += ['--wait']
-        else:
-            self._listen()
-            argv += ['--client']
+    def _get_argv_for_launch(self):
+        argv = [sys.executable]
+        argv += [ptvsd.__main__.__file__]
+        argv += ['--client']
         argv += ['--host', 'localhost', '--port', str(self.ptvsd_port)]
+        if self.multiprocess_port_range:
```
Comment (int19h) [inline pytests/helpers/session.py]: We don't have this in the product at the moment (it was there for a while as an artifact of an older multiproc implementation, but it's not needed for the current one).

File: pytests/helpers/session.py [hunk 4]
```diff
@@ -178,10 +184,41 @@ def prepare_to_run(self):
...
+        print('Preparing debug session with method %r' % str(self.start_method))
+        argv = []
+        if self.start_method == START_METHOD_LAUNCH:
+            self._listen()
+            argv += self._get_argv_for_launch()
+        elif self.start_method == START_METHOD_CMDLINE:
+            argv += self._get_argv_for_attach_using_cmdline()
+        elif self.start_method == START_METHOD_IMPORT:
+            argv += self._get_argv_for_attach_using_import()
+            ptvsd_path = os.path.dirname(os.path.dirname(ptvsd.__main__.__file__))
```
Comment (int19h) [inline pytests/helpers/session.py]: Can you add a TODO comment here saying that we need to get rid of this after we fully switch to Tox? With Tox, ptvsd will be installed in the test environment as a proper package, so no PYTHONPATH patching should ever be necessary).

File: pytests/func/test_breakpoints.py [hunk 5]
```diff
@@ -8,16 +8,16 @@
...
+from pytests.helpers.timeline import Event, Response
 from pytests.helpers.pathutils import get_test_root, compare_path
-from pytests.helpers.session import START_METHOD_LAUNCH, START_METHOD_CMDLINE
+from pytests.helpers.session import START_METHOD_LAUNCH, START_METHOD_CMDLINE, START_METHOD_IMPORT
 
 BP_TEST_ROOT = get_test_root('bp')
 
 
-@pytest.mark.parametrize('start_method', [START_METHOD_LAUNCH, START_METHOD_CMDLINE])
+@pytest.mark.parametrize('start_method', [START_METHOD_LAUNCH, START_METHOD_CMDLINE, START_METHOD_IMPORT])
```
Comment (int19h) [inline pytests/func/test_breakpoints.py]: This is going to be the new default for all tests, right? (i.e. unless a test has a good reason to exclude some method, it should cover all of them). If so, let's make the `start_method` fixture return this by default, and drop `@parametrize` on all tests that use the default.

Comment (karthiknadig) [reply pytests/func/test_breakpoints.py]: Agreed. Will make the change.

**Comment by karthiknadig** [comment]
This looks like it keeps hitting the NoneType issue when importing ptvsd. I can't merge this until that issue is fixed.

**Comment by int19h** [comment]
I was actually hoping to use this to repro it! Can you disable that start method for all tests, but merge the rest?

**Pull request merged**